### ISSUE TITLE
BuilderParticipant.getDerivedResources has a NxM issue

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#225]
`org.eclipse.xtext.builder.BuilderParticipant.getDerivedResources(Delta, Map<String, OutputConfiguration>, Map<OutputConfiguration, Iterable<IMarker>>)` has a NxM issue

if you do a full build on a project that contains xtext-based models and there generated files and do not do a clean build before then `BuilderParticipant.getDerivedResources` will have a NxM runtime

(x deltas times x markers/generatedfiles - i assue there is one foreach source file)

**Comment by cdietrich** [comment]
`org.eclipse.core.resources.IMarker.getAttribute(String)`
goes into `org.eclipse.core.internal.resources.MarkerAttributeMap.get(Object)`
which call `String.intern`.
this might become really expensive for 10000(Deltas) times 10000(Resources) = 100.000.000 calls of `org.eclipse.xtext.builder.DerivedResourceMarkers.getSource(IMarker)`

**Comment by cdietrich** [comment]
in a normal clean/build cycle the derived resources are delted by org.eclipse.xtext.builder.BuilderParticipant.cleanOutput(IBuildContext, OutputConfiguration, EclipseResourceFileSystemAccess2, IProgressMonitor) so this problem does not occur.

**Comment by cdietrich** [comment]
trace from yourkit

```
Stacks at 2017-04-04 04:31:48 PM. Uptime is 2m 47s 200ms.

Worker-7 State: RUNNABLE CPU usage on sample: 982ms
java.lang.String.intern() String.java (native)
org.eclipse.core.internal.resources.MarkerAttributeMap.get(Object) MarkerAttributeMap.java:135
org.eclipse.core.internal.resources.MarkerInfo.getAttribute(String) MarkerInfo.java:108
org.eclipse.core.internal.resources.Marker.getAttribute(String) Marker.java:99
org.eclipse.xtext.builder.DerivedResourceMarkers.getSource(IMarker) DerivedResourceMarkers.java:164
org.eclipse.xtext.builder.BuilderParticipant.getDerivedResources(IResourceDescription$Delta, Map, Map) BuilderParticipant.java:390
org.eclipse.xtext.builder.BuilderParticipant.doBuild(List, Map, Map, IXtextBuilderParticipant$IBuildContext, EclipseResourceFileSystemAccess2, IProgressMonitor) BuilderParticipant.java:260
org.eclipse.xtext.builder.BuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) BuilderParticipant.java:221
com.advantest.itee.stpl.internal.builder.OutputFolderCleaningBuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) OutputFolderCleaningBuilderParticipant.java:56
com.advantest.itee.stpl.internal.builder.DelegatingXtextBuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) DelegatingXtextBuilderParticipant.java:46
org.eclipse.xtext.builder.impl.RegistryBuilderParticipant$DeferredBuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) RegistryBuilderParticipant.java:161
org.eclipse.xtext.builder.impl.RegistryBuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) RegistryBuilderParticipant.java:69
com.advantest.itee.tpl.ui.internal.ExtendedRegistryBuilderParticipant.build(IXtextBuilderParticipant$IBuildContext, IProgressMonitor) ExtendedRegistryBuilderParticipant.java:99
org.eclipse.xtext.builder.impl.XtextBuilder.doBuild(ToBeBuilt, IProgressMonitor, IXtextBuilderParticipant$BuildType) XtextBuilder.java:252
com.advantest.itee.tpl.ui.internal.GlobalSharedXTextBuilder.doBuild(ToBeBuilt, IProgressMonitor, IXtextBuilderParticipant$BuildType) GlobalSharedXTextBuilder.java:261
org.eclipse.xtext.builder.impl.XtextBuilder.fullBuild(IProgressMonitor, boolean) XtextBuilder.java:280
org.eclipse.xtext.builder.impl.XtextBuilder.build(int, Map, IProgressMonitor) XtextBuilder.java:117
com.advantest.itee.tpl.ui.internal.GlobalSharedXTextBuilder.build(int, Map, IProgressMonitor) GlobalSharedXTextBuilder.java:151
org.eclipse.core.internal.events.BuildManager$2.run() BuildManager.java:726
org.eclipse.core.runtime.SafeRunner.run(ISafeRunnable) SafeRunner.java:42
org.eclipse.core.internal.events.BuildManager.basicBuild(int, IncrementalProjectBuilder, Map, MultiStatus, IProgressMonitor) BuildManager.java:199
org.eclipse.core.internal.events.BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, ICommand[], MultiStatus, IProgressMonitor) BuildManager.java:239
org.eclipse.core.internal.events.BuildManager$1.run() BuildManager.java:292
org.eclipse.core.runtime.SafeRunner.run(ISafeRunnable) SafeRunner.java:42
org.eclipse.core.internal.events.BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, MultiStatus, IProgressMonitor) BuildManager.java:295
org.eclipse.core.internal.events.BuildManager.basicBuildLoop(IBuildConfiguration[], IBuildConfiguration[], int, MultiStatus, IProgressMonitor) BuildManager.java:351
org.eclipse.core.internal.events.BuildManager.build(IBuildConfiguration[], IBuildConfiguration[], int, IProgressMonitor) BuildManager.java:374
org.eclipse.core.internal.events.AutoBuildJob.doBuild(IProgressMonitor) AutoBuildJob.java:143
org.eclipse.core.internal.events.AutoBuildJob.run(IProgressMonitor) AutoBuildJob.java:241
org.eclipse.core.internal.jobs.Worker.run() Worker.java:54

``` 

screenshot from yourkit

![yourkit](https://cloud.githubusercontent.com/assets/238736/24664543/38db91de-195b-11e7-8f09-0c580fb70997.png)

**Comment by cdietrich** [comment]
Proposed solution:

wrap result of `org.eclipse.xtext.builder.BuilderParticipant.getGeneratorMarkers(IProject, Collection<OutputConfiguration>)` into a `ForwardingMap`/Delegate and do build a reverse lookup map inside this class. then inside `org.eclipse.xtext.builder.BuilderParticipant.getDerivedResources(Delta, Map<String, OutputConfiguration>, Map<OutputConfiguration, Iterable<IMarker>>)` do an instanceof on the Wrapper Type and ask the Wrapper for the reverse Lookup Value

**Comment by cdietrich** [comment]
Performance improvement (BuilderParticipant.doBuild) on a 2000 xtend files project (touch all): 1:4
Performance improvement (BuilderParticipant.doBuild) on a 5000 xtend files project (touch all): > 1:8

**Comment by cdietrich** [comment]
fixed in Xtext 2.12
